### ISSUE TITLE
Problems deploying to VirtualBox

### ISSUE THREAD

**Issue by gilligan** [NixOS/nixops#473]
I was trying to follow the setup described in the nixos manual on my OSX machine. The deployment fails because nixops fails to connect via ssh. The following is pretty much the whole process:

```
$ ifconfig vboxnet0
vboxnet0: flags=8842<BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	ether 0a:00:27:00:00:00
$ nixops list
$ nixops create 
$ nixops create ./trivial.nix ./trivial-vbox.nix -d trivial
created deployment ‘1226e00f-4cf7-11e6-bd86-6c400892f666’
1226e00f-4cf7-11e6-bd86-6c400892f666
% nixops info -d trivial
Network name: trivial
Network UUID: 1226e00f-4cf7-11e6-bd86-6c400892f666
Network description: Web server
Nix expressions: /Users/tpflug/Development/nixops-playground/trivial.nix /Users/tpflug/Development/nixops-playground/trivial-vbox.nix

+-----------+---------------+------------+-------------+------------+
| Name      |     Status    | Type       | Resource Id | IP address |
+-----------+---------------+------------+-------------+------------+
| webserver | Missing / New | virtualbox |             |            |
+-----------+---------------+------------+-------------+------------+
$ nixops deploy -d trivial
$ nixops deploy -d trivial
webserver> creating VirtualBox VM...
webserver> Virtual machine 'nixops-1226e00f-4cf7-11e6-bd86-6c400892f666-webserver' is created and registered.
webserver> UUID: 610c9e37-b24a-42f5-94ac-1f94dddeb643
webserver> Settings file: '/Users/tpflug/VirtualBox VMs/nixops-1226e00f-4cf7-11e6-bd86-6c400892f666-webserver/nixops-1226e00f-4cf7-11e6-bd86-6c400892f666-webserver.vbox'
webserver> creating disk ‘disk1’...
webserver> 0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
webserver> Clone medium created in format 'VDI'. UUID: 3cf09760-435f-4c75-9751-0737b036fedd
webserver> attaching disk ‘disk1’...
webserver> Waiting for VM "nixops-1226e00f-4cf7-11e6-bd86-6c400892f666-webserver" to power on...
webserver> VM "nixops-1226e00f-4cf7-11e6-bd86-6c400892f666-webserver" has been successfully started.
webserver> waiting for IP address........................... 169.254.215.18

ssh: connect to host 169.254.215.18 port 22: Operation timed out
webserver> could not connect to ‘root@169.254.215.18’, retrying in 1 seconds...
```

It seems weird to me that the vboxnet0 device would have a 192.168.x.x ip address whereas nixops tries to ssh to 169.254.x.x - Does that actually make sense ?

**Update**: With DHCP server disabled for vboxnet0 the deployment fails because nixops fails to ssh in. Once the DHCP server has been enabled/configured it fails later on during deployment. From the log on the virtual machine:

```
Jul 19 07:50:36 webserver systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start timed out.
Jul 19 07:50:36 webserver systemd[1]: Timed out waiting for device dev-vboxguest.device.
Jul 19 07:50:36 webserver systemd[1]: Dependency failed for Get NixOps SSH Key.
Jul 19 07:50:36 webserver systemd[1]: get-vbox-nixops-client-key.service: Job get-vbox-nixops-client-key.service/start failed with result 'dependency'.
Jul 19 07:50:36 webserver systemd[1]: Dependency failed for VirtualBox Guest Services.
Jul 19 07:50:36 webserver systemd[1]: virtualbox.service: Job virtualbox.service/start failed with result 'dependency'.
Jul 19 07:50:36 webserver systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start failed with result 'timeout'.
Jul 19 07:50:36 webserver nixos[1285]: switching to system configuration /nix/store/lpbfbm03hs2kii2rv23nwayh3fkg12hl-nixos-system-webserver-16.09.git.963be93 failed (status 4)
```

A consecutive call succeeds:
```
$ nixops deploy -d trivial                                                                                                                                                    
building all machine configurations...
trace: warning: The option `boot.loader.grub.timeout' defined in `/nix/store/80dcak353g73kpar379s9y5q1wyak8ml-nixops-1.3.1/share/nix/nixops/virtualbox-image-nixops.nix' has been renamed to `boot.loader.timeout'.
webserver> copying closure...
webserver> updating GRUB 2 menu...
webserver> activating the configuration...
webserver> setting up /etc...
webserver> the following new units were started: get-vbox-nixops-client-key.service, sys-devices-pci0000:00-0000:00:03.0-virtio0-net-eth0.device, sys-devices-pci0000:00-0000:00:08.0-virtio1-net-eth1.device, sys-devices-pci0000:00-0000:00:0d.0-ata1-host0-target0:0:0-0:0:0:0-block-sda-sda1.device, sys-devices-pci0000:00-0000:00:0d.0-ata1-host0-target0:0:0-0:0:0:0-block-sda.device, sys-devices-platform-serial8250-tty-ttyS0.device, sys-devices-platform-serial8250-tty-ttyS1.device, sys-devices-platform-serial8250-tty-ttyS2.device, sys-devices-platform-serial8250-tty-ttyS3.device, sys-devices-virtual-input-mice.device, sys-devices-virtual-mem-random.device, sys-devices-virtual-misc-rfkill.device, sys-devices-virtual-misc-vboxguest.device, sys-subsystem-net-devices-enp0s3.device, sys-subsystem-net-devices-enp0s8.device, virtualbox.service
webserver> activation finished successfully
trivial> deployment finished successfully
```

----

**nixops**: NixOps 1.3.1
**OSX/nixpkgs**: 16.09pre-git
**nixos/nixpkgs**: 16.09pre85931.125fff"

So I tried this under OSX and on my nixos laptop. Same scenario

**Comment by gilligan** [comment]
This gist also has the `trivial` nix files: https://gist.github.com/gilligan/c36c13419c9a94894e4f7fbe2acc203d

**Comment by gilligan** [comment]
I just changed the vboxnet0 address to `169.254.215.1` and now it works. This looks like it would be a problem.

**Comment by gilligan** [comment]
I just tried with an older version of VirtualBox (5.0.2r102096) - it does not work any better.

**Comment by gilligan** [comment]
Might be related to https://github.com/NixOS/nixops/issues/332

**Comment by gilligan** [comment]
Just tried with VirtualBox 4.3.x - no luck.

**Comment by gilligan** [comment]
It seems what was required for this to work was that the DHCP server is enabled for vboxnet0 - Once I enabled that the deployment proceeded. It did however fail. I logged into the machine and looked at the log output:

```
Jul 19 07:09:12 webserver systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/sta
Jul 19 07:09:12 webserver systemd[1]: Timed out waiting for device dev-vboxguest.device.
Jul 19 07:09:12 webserver systemd[1]: Dependency failed for VirtualBox Guest Services.
Jul 19 07:09:12 webserver systemd[1]: virtualbox.service: Job virtualbox.service/start f
Jul 19 07:09:12 webserver systemd[1]: Dependency failed for Get NixOps SSH Key.
Jul 19 07:09:12 webserver systemd[1]: get-vbox-nixops-client-key.service: Job get-vbox-n
Jul 19 07:09:12 webserver systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/sta
Jul 19 07:09:12 webserver nixos[2077]: switching to system configuration /nix/store/m0mk
Jul 19 07:09:12 webserver sshd[714]: Received disconnect from 192.168.56.1: 11: disconne
Jul 19 07:09:12 webserver sshd[714]: Disconnected from 192.168.56.1
Jul 19 07:09:12 webserver sshd[714]: pam_unix(sshd:session): session closed for user root
```

When executing the deployment for a second time after that it succeeded ;/

**Comment by gilligan** [comment]
I just tried on NixOS and I am having the exact same issues - The first deployment will fail and on a second run it will report success.

**Comment by gilligan** [comment]
I tried with nixops from master - it still needs a second deployment to succeed.

**Comment by domenkozar** [comment]
When you deploy from scratch?

**Comment by gilligan** [comment]
@domenkozar I am actually having a hard time keeping an overview. I have tried a *lot* of things/ combinations/ variations at this point but nothing I tried worked as expected.

Incomplete summary:

* Just having `vboxnet0` as host-only networking device does not suffice. In order for nixops to even succeed to ssh into the virtual machine a dhcp server needs to be enabled on vboxnet0 - this can be enabled in the second tab of the VirtualBox networking settings of host-only networking for vboxnet0. I *think* that previous versions of VirtualBox might have this enabled by default. It is not anymore now.

* With vboxnet0 running with dhcp the deployment can at least start. It will however fail with what seems to be the problem described in #391 - Unrelated to apache but rather about vbox guest additions service. Note that even if nixops says the deployment failed the webserver configured in trivial.nix is actually reachable - Running deploy a second time yields a successful deployment.

* nixops from master did not solve this either. I think the problem changed somewhat because guest additions did not fail to come up but nixops also reported the deployment as failed. Again deploying a second time "fixes" this. I can run this again and provide log output from the deployment.

I tried all kinds of combinations at this point (osx/nixos, different nixops versions, different nixpkgs revisions) but I couldn't find any combination in which I could deploy the simple `trivial.nix` example with success at first try.

**Comment by gilligan** [comment]
@domenkozar just did a deployment with nixops from master

```
[root@jenkinsMaster:~]# systemctl status dev-vboxguest.device
● dev-vboxguest.device
   Loaded: loaded
   Active: inactive (dead)

Jul 19 12:47:11 nixos systemd[1]: Found device /dev/vboxguest.
Jul 19 12:50:33 jenkinsMaster systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start timed out.
Jul 19 12:50:33 jenkinsMaster systemd[1]: Timed out waiting for device dev-vboxguest.device.
Jul 19 12:50:33 jenkinsMaster systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start failed with result 'timeout'.

[root@jenkinsMaster:~]# udevadm info -n /dev/vboxguest
P: /devices/virtual/misc/vboxguest
N: vboxguest
E: DEVNAME=/dev/vboxguest
E: DEVPATH=/devices/virtual/misc/vboxguest
E: MAJOR=10
E: MINOR=58
E: PATH=/nix/store/g2rycyddhwzb9dhbb10pqyxx8nrrhfnz-udev-path/bin:/nix/store/g2rycyddhwzb9dhbb10pqyxx8nrrhfnz-udev-path/sbin
E: SUBSYSTEM=misc
E: TAGS=:systemd:
E: USEC_INITIALIZED=5863


[root@jenkinsMaster:~]# journalctl -b | grep vboxguest
Jul 19 12:47:10 nixos kernel: vboxguest: misc device minor 58, IRQ 20, I/O port d020, MMIO at 00000000f0000000 (size 0x400000)
Jul 19 12:47:10 nixos kernel: vboxguest: Successfully loaded version 5.0.4 (interface 0x00010004)
Jul 19 12:47:10 nixos systemd[1]: Expecting device dev-vboxguest.device...
Jul 19 12:47:11 nixos systemd[1]: Found device /dev/vboxguest.
Jul 19 12:50:33 jenkinsMaster systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start timed out.
Jul 19 12:50:33 jenkinsMaster systemd[1]: Timed out waiting for device dev-vboxguest.device.
Jul 19 12:50:33 jenkinsMaster systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start failed with result 'timeout'.
```

So I guess this is identical to what you described in #391

**Comment by domenkozar** [comment]
Again, did you deploy to an existing vm or from scratch?

**Comment by gilligan** [comment]
@domenkozar ah sorry - from scratch.

**Comment by gilligan** [comment]
@domenkozar can you successfully deploy something to virtualbox? if so what is your setup? Also what other information can I provide so that we can nail down what is happening here?

**Comment by domenkozar** [comment]
So the first issue is a duplicate of https://github.com/NixOS/nixops/issues/411, I can reproduce the second issue and I'll check what's going on really.

**Comment by domenkozar** [comment]
@gilligan the second issue should be fixed once new image is generated for 16.03. See https://github.com/NixOS/nixpkgs/commit/5cc7bcda3023b01bd926eccd0c5f095a050c5ab0#commitcomment-18322469 what's blocking that.

**Comment by danbst** [comment]
So, is there a workaround for this? I'm getting this too
